### ISSUE TITLE
Proposal: Iteration expansion

### ISSUE THREAD

**Issue by jonathandturner** [nushell/nushell#1120]
**Is your feature request related to a problem? Please describe.**

Currently, we have two styles of command that work over the pipeline: per-item and whole-stream.

Per-item does some amount of work on each item of the stream and is expected to be called on each item individually. It may return 0-or-more results in a stream that becomes part of the pipeline.

Whole-stream runs once and is expected to continue running while there is a stream.

We also allow commands to receive an $it-path as an argument. Commands must know how to handle this properly and switch modes into a per-item style to update $it for each item in the stream.

There are some gotchas in the current implementation of this:

* Commands have to declare ahead of time whether they are per-item or whole-stream. This makes it difficult when a command is sometimes to be treated per-item (eg, you use $it as a parameter) and sometimes as whole-stream (eg, you want to do some work even when $it is not a parameter).
* Each command has to handle $it-expansion, making implementation error prone and ad hoc.
* $it-expansion in externals is currently via string replacement and doesn't work the same way as internal commands
* It's currently unclear what it means to use $it: do you intend to have both an argument and a stream input or *just* the argument?

**Describe the solution you'd like**

I propose we add more formalized expansion steps before evaluation of a pipeline. In this proposal, we introduce "iteration expansion" as one of those expansion steps.

Iteration expansion works like this:

## Detection
* A command's parameters are checked to see if they use $it-paths. These must be at the top-level of the argument and not inside a block
* If detected, the command is iteration-expanded

## Expansion
* A command that's iteration expanded is expanded from `command arg1 arg2` form to: `each {command arg1 arg2}` where `each` is a whole-stream command that will invoke the block on each iteration

## The `each` command
* `each` invokes the block once per item in the pipeline
* `each` does *not* pass input into its block, it only sets the value of $it during the evaluation of the block
* `each` outputs 0-or-more items into the stream as a result of evaluating the block

## Notes
* iteration-expansion works over both internal and external commands (see examples below)
* iteration-expansion also works with `where`-expansion because where-expansion turns the condition into a block, which is not iteration-expanded
* This effectively removes per-item commands altogether, conceptually unifying all commands as whole-stream
  * This is with the current exception of plugins, which are run from a whole-stream plugin harness in a per-item fashion. There are currently issues for plugins that want to use $it.  With iteration-expansion, we remove some of the unnecessary workarounds and allow plugins to more directly take advantage of $it as a parameter.  (side note: we can further improve the plugin protocol to make this more efficient in the future)

**Describe alternatives you've considered**
We could require all iteration to be explicit, but, as with `where`, a clear, understandable sugar can help write-ability.

### Internal
Before expansion:
`ls | echo $it.name`

After expansion:
`ls | each { echo $it.name }`

### External
Before expansion:
`ls | get name | ^echo $it`

After expansion:
`ls | get name | each { ^echo $it }`

**Comment by thegedge** [comment]
This could be related to https://github.com/nushell/nushell/issues/1309#issuecomment-580444134

> Perhaps we can go for a quick fix to get things working, but ideally we could make internals feel like externals by having PerItem ($it arguments) and WholeStream (no $it args, feed into stdin) spawning.

I think there are three things we may want to support:

1. whole stream, pipe stdin into command `... | external <args but no $it>`
2. per item, $it as arg `... | external <args with $it>`
3. per item, $it as stdin `... | ???`

My thought is that (3) is not super common, so it could be written out as something like this:

```
... | each { echo <something with $it> | external <other args> }
```

**Comment by jonathandturner** [comment]
partially implemented by #1668

**Comment by jonathandturner** [comment]
Finished with #1682
